docs/code-generation-design.md

```diff
@@ -0,0 +1,251 @@
+* A language-specific method to map enumerated value names - this function should handle the means by which a enumeration or identity name (as specified by the RFC7950 grammar) can be safely translated to a value which can be used in the target language.
+
+An implementation of a `LangMapper` keeps its own internal state that is not made accessible to the `ygen` library. Since a `LangMapper` implementation has access to the `yang.Entry`, it is able to use the entire context of a particular YANG entry to be able to choose how to map it to its name in the generated language.
+
+Other than the `LangMapper` interface, no other language-specific mapping code is implemented in the production of the IR.
+
+### Defining the ygen IR
+
+Currently, the `ypathgen` library defines a `GetDirectoriesAndLeafTypes` method that can be called to return:
+
+ * A map of the directories that are used in the generated code, keyed by the YANG path (represented as a string).
+ * A map of maps, keyed by directory name, with the inner map keyed by field name, returning the language-type to be used for a particular leaf.
+
+These types are sufficient for `ypathgen` to generate the code that it currently generates, but insufficient for all languages to be produced (or other formats of Go). Equally, there is some duplication between these types that requires cross-referencing between them which could be simplified.
+
+We propose to modify this return format - and define the IR to be broken down into two subsets - encapsulated by a common type:
+
+```
+type Definitions struct {
+	// ParsedTree is the set of parsed directory entries, keyed by the YANG path
+	// to the directory.
+	ParsedTree map[string]*ParsedDirectory
+	
+	// Enums is the set of enumerated entries that are extracted from the generated
+	// code. The key of the map is the global name of the enumeration.
+	Enums map[string]*EnumeratedYANGType
+
+	// ModelData stores the metadata extracted from the input YANG modules.
+	ModelData     []*gpb.ModelData
+}
+```
+
+#### Definitions for Directory / Leaf Nodes
+
+The proposed IR for the directory types in the returned code is as follows. The base philosophy of whether fields are included within this representation is to keep to the set of fields made available to the code output library to their minimum, to ensure that there is clear encapsulation of the complexities of the YANG hierarchy within `ygen`.
+
+```golang
+// ParsedDirectory describes a node that is the root of a subtree
```

Comment:
nit: "a node that is the root of a subtree" -> "an internal node"

Reply:
Fixed, thanks.